### ISSUE TITLE
RAII Proposal

### ISSUE THREAD

**Issue by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#466]
In its current state, the game has lots of uninitialized variables and fields all over the place. And there are multiple places in the code that depend on these uninitialized values to make decisions.

We need to fix that. I would like to see us fix it as part of 0.9.x.

c++11 lets you initialize class fields on the same line where they are declared, which I guess you couldn't do in previous versions of the standard. c++11 also introduces brace-initializers (`{}`). From [what I've been reading](https://www.informit.com/articles/article.aspx?p=1852519), brace-initializers are a consistent syntax for initializing fields and variables with all four types of possible values, each of which had a different syntax in previous versions of C++.

So I propose that we use brace-initializers `{}` to initialize every variable and every field, right where it's declared, wherever that makes sense across the Vega Strike codebase. If we adopt this convention, it should make it easy to spot places in the code that aren't up to the RAII standard yet, and fix them.

There are some places in the code that aren't ready to be implemented this way yet. Such as `Unit`s, and their complicated system of references and backreferences that gets dismantled somehow after a Unit dies. We can perhaps leave these parts of the code for later.

@BenjamenMeyer has said that he'd rather not have the compiler generate implicit constructors for our classes; he would prefer all constructors to be explicitly defined. Well, could we have **both** explicit constructors **and** empty brace-initializers at the point where fields are declared? We would end up double-initializing many fields. But at least we'd be sure that they **were** initialized at least once -- unlike with the current situation.

What do you guys think?

**Comment by stephengtuggy** [comment]
Relates to #346

**Comment by BenjamenMeyer** [comment]
Yes, we need to move away from having uninitialized values. I'd need to study the code a little more to see if I agree on the method though.

Classes and structs have always been able to be initialized from the get-go - just use the class/struct constructor.
Also, some things you can't easily initialize - like std::map() so then it's better to have a function that manages the initialization:
```
static std::map<std::string, bool> foo;
bool accessFoo(s std::string) {
   static bool fooInitialized = false;
   if !fooInitialized {
       // load foo values here
       fooInitialized = true;
   }

   bool return_value = false;
   // get actual value here
   return return_value;
}
```
So really it comes down to the type in question as to the appropriate method of initialization.

**Comment by BenjamenMeyer** [comment]
we've started using `{}` initializers; so I think we can close this unless @stephengtuggy there's another reason to keep it open?

**Comment by stephengtuggy** [comment]
@BenjamenMeyer sure. Closing
